src/rp2/balance.py

```diff
@@ -119,53 +118,60 @@ def __init__(
         from_account: Account
         to_account: Account
 
-        # Balances for bought and earned currency
-        for transaction in self.__input_data.unfiltered_in_transaction_set:
-            if transaction.timestamp.date() > to_date:
-                break
-            in_transaction: InTransaction = cast(InTransaction, transaction)
-            to_account = Account(in_transaction.exchange, in_transaction.holder)
-            acquired_balances[to_account] = acquired_balances.get(to_account, ZERO) + in_transaction.crypto_in
-            final_balances[to_account] = final_balances.get(to_account, ZERO) + in_transaction.crypto_in
+        in_transactions: List[InTransaction] = cast(List[InTransaction], list(self.__input_data.unfiltered_in_transaction_set))
+        intra_transactions: List[IntraTransaction] = cast(List[IntraTransaction], list(self.__input_data.unfiltered_intra_transaction_set))
+        out_transactions: List[OutTransaction] = cast(List[OutTransaction], list(self.__input_data.unfiltered_out_transaction_set))
 
-        # Balances for currency that is moved across accounts
-        for transaction in self.__input_data.unfiltered_intra_transaction_set:
-            if transaction.timestamp.date() > to_date:
-                break
-            intra_transaction: IntraTransaction = cast(IntraTransaction, transaction)
-            from_account = Account(intra_transaction.from_exchange, intra_transaction.from_holder)
-            to_account = Account(intra_transaction.to_exchange, intra_transaction.to_holder)
-            sent_balances[from_account] = sent_balances.get(from_account, ZERO) + intra_transaction.crypto_sent
-            received_balances[to_account] = received_balances.get(to_account, ZERO) + intra_transaction.crypto_received
-            final_balances[from_account] = final_balances.get(from_account, ZERO) - intra_transaction.crypto_sent
-            final_balances[to_account] = final_balances.get(to_account, ZERO) + intra_transaction.crypto_received
-            if (
-                not RP2Decimal.is_equal_within_precision(final_balances[from_account], ZERO, CRYPTO_BALANCE_DECIMAL_MASK)
-                and final_balances[from_account] < ZERO
-                and not configuration.allow_negative_balances
-            ):
-                raise RP2ValueError(
-                    f'{intra_transaction.asset} balance of account "{from_account.exchange}" (holder "{from_account.holder}") went negative '
-                    f'({final_balances[from_account]}) on the following transaction: {intra_transaction}'
-                )
-
-        # Balances for sold and gifted currency
-        for transaction in self.__input_data.unfiltered_out_transaction_set:
+        transactions: List[Union[InTransaction, IntraTransaction, OutTransaction]] = in_transactions + intra_transactions + out_transactions
```

Comment:
Could this be `List[AbstractTransaction]` instead of a `Union`?

Reply:
Solved this in the most recent commit via this approach: https://peps.python.org/pep-0484/#type-variables-with-an-upper-bound.

This can be simplified when 3.11 is the minimum supported version via PEP 673